### DISCUSSION TITLE
Give option to 'Step back to closure'

### DISCUSSION THREAD

**Discussion by bgirard** [RecordReplay/devtools#5897] [Ideas]
There's some simple case where you can set a breakpoint in the closure context and just step backwards. This works fine closure execution roughly is 1:1 with parent/child. But in some scenarios they are not and this might be scenarios where you run into other limitations like too many breakpoint on this line. Some examples:
* Webpack module registration might register 5,000 modules in one order, and run them in another. So if you fail to run one, it's very hard to jump back to the closure that registered the module to see what went run.
* React hooks. Let's say you have many instance of one component that get rendered and might be long live. Some interesting effect fires much later, like some mutation observer. You might want to jump back to the render of THAT component INSTANCE. Which would be the closure parent.

Ideally we'd add a new button in the stepping UI call 'Set back to closure' (naming is hard, find a better name :D). In the VM there's a shared closure object, so you roughly want to step back to when that was allocated if that's feasible with the Replay platform.

**Comment by bgirard** [comment]
If we had the option to step back to when an object was created I bet we wouldn't strictly need this UI. Although I bet when you support that one, you can easily support this one.

**Reply by jasonLaster** [reply]
re watchpoints: I think you're right to say this is a special case of https://github.com/RecordReplay/devtools/discussions/5495

Curious what @bhackett1024 thinks.

**Reply by bhackett1024** [reply]
I agree with the above, this feels like it needs watchpoint data and (possibly) a watchpoint based UI, where it's easy to see a timeline of the history of a particular object and jump to different points on it.  From the VM's perspective, the only thing linking the inner frame with the place that created it is the callee function object.  There *might* be a call object holding variables in the outer function's scope that are still accessible, but that isn't guaranteed.  If we can create a timeline for the callee function object we can find the frame which created it, but once we're able to do that we might as well do the same thing for all the objects the program creates.
